std/range/package.d

```diff
@@ -4688,7 +4688,13 @@ if ((isIntegral!(CommonType!(B, E)) || isPointer!(CommonType!(B, E)))
         && isIntegral!S)
 in
 {
-    assert(!(step == 0 && begin != end));
+    import std.conv : unsigned;
+
+    assert(step != 0 || begin == end);
+    if (step > 0)
+        assert(unsigned((end - begin) / step) <= size_t.max);
+    else if (step < 0)
+        assert(unsigned((end - begin) / -step) <= size_t.max);
```

Comment:
Oh, you're right... it should be `(end - begin) / step`, since `end - begin` will be negative, but `step` is also already negative.  So actually, there's no need for two branches in the code at all; it can be just:
````
if (step != 0)
    assert(unsigned((end - begin) / step) <= size_t.max);
````
Except you might have to be careful about what happens when `begin` is `int.max` and `end` is `int.min`...